### PR TITLE
Test that AsyncLocal<class> changed event is not raised multiple times for one change in value (#29316)

### PR DESCRIPTION
**Pull request by kouvel**
Port of https://github.com/dotnet/corefx/pull/29316 to 2.2

- Depends on and tests for https://github.com/dotnet/coreclr/pull/19607
- Tests for https://github.com/dotnet/coreclr/issues/17758

### CODE REVIEW

File: src/System.Threading/tests/AsyncLocalTests.cs [hunk 1]
```diff
@@ -153,6 +153,56 @@ public static async Task NotifyOnValuePropertyChange()
             return;
         }
 
+        [Fact]
+        public static async Task NotifyCountOnObjectValueChange()
+        {
```
Comment (tarekgh) [inline src/System.Threading/tests/AsyncLocalTests.cs]: The test looks good but I am wondering if we have tests setting the asyncLocal values to null too? testing null values with and without opting in to the notification would be nice.

Comment (kouvel) [reply src/System.Threading/tests/AsyncLocalTests.cs]: It looks like this test does a bit of that:
https://github.com/dotnet/corefx/blob/eceff3b84d3a6e1865aab09f3ee9a4be9ae4d5ce/src/System.Threading/tests/AsyncLocalTests.cs#L517

Comment (tarekgh) [reply src/System.Threading/tests/AsyncLocalTests.cs]: does this test used when creating using ```AsyncLocal(Action<AsyncLocalValueChangedArgs<T>> valueChangedHandler)``` and when using ```AsyncLocal()``` constructors?

Comment (kouvel) [reply src/System.Threading/tests/AsyncLocalTests.cs]: No only the empty constructor, perhaps it can be updated to test both, I'll consider for master

**Comment by kouvel** [comment]
Closing, see https://github.com/dotnet/coreclr/pull/19607

**Pull request closed without merging**